Using Keycloak as an External Identity Provider for Eclipse Che on Op… (#90)
eclipse-che/blog: _posts/2026-03-31-openshift-external-idp.adoc

---
title: Using Keycloak as an External Identity Provider for Eclipse Che on OpenShift
layout: post
author: Anatolii Bazko
description: >-
  Install Eclipse Che on OpenShift with Keycloak as an external OIDC identity provider.
categories: []
keywords: ['OpenShift','Keycloak','OIDC','identity provider','authentication','Eclipse Che']
slug: /@tolusha/openshift-external-idp
---

Starting with OpenShift 4.20, administrators can configure
link:https://docs.redhat.com/en/documentation/openshift_container_platform/4.20/html/authentication_and_authorization/external-auth[external OIDC identity providers]

> for direct authentication to bypasses the built-in OAuth server and uses the

Reviewer: ```suggestion
for direct authentication to bypass the built-in OAuth server and use the
```

external identity provider directly.
This enables direct integration with an external OIDC provider,

> allowing you to leverage advanced capabilities beyond the built-in OAuth server.

Reviewer 2: "leverage" → "use"

Suggestion:
```
allowing you to use advanced capabilities beyond the built-in OAuth server.
```

"leverage" is on the Red Hat/IBM Style complex-word list. "use" is direct and clear.

> allowing you to leverage its advanced capabilities beyond the built-in OAuth server.

Reviewer 3: ```suggestion
allowing you to leverage advanced capabilities beyond the built-in OAuth server.
```

Eclipse Che now fully supports this setup: you can deploy Che on an OpenShift cluster

> that authenticates users, for example, through an external Keycloak instance.

Reviewer 3: ```suggestion
that authenticates users through an external Keycloak instance, for example.
```

In this post, we walk through the end-to-end configuration.

== Step-by-Step Configuration

This guide assumes that Keycloak is already deployed and configured as
an external identity provider for OpenShift.

=== 1. Create the Eclipse Che Client in Keycloak

In the Keycloak Admin Console, create a new client for Eclipse Che:

. Navigate to your OpenShift realm and select *Clients*.
. Click *Create client*.
. Set the *Client ID* to `eclipse-che`.
. On the *Capability config* page, toggle *Client authentication* to *On*.
. Save the client and copy the *Client secret* from the *Credentials* tab.

> NOTE: The redirect URI will be configured later, after Eclipse Che is deployed.

Reviewer 3: ```suggestion
NOTE: The redirect URI will be configured after Eclipse Che is deployed.
```

        CHE_OIDC_GROUPS__PREFIX: '<GROUPS_PREFIX>'
        CHE_OIDC_USERNAME__CLAIM: '<USERNAME_CLAIM>'
        CHE_OIDC_USERNAME__PREFIX: '<USERNAME_PREFIX>'
EOF
----

=== 7. Deploy Eclipse Che

[source,shell]
----
chectl server:deploy --platform openshift --che-operator-cr-patch-yaml che-patch.yaml
----

image::/assets/img/openshift-external-idp/eclipse-che-deployed.png[Eclipse Che deployed]
{nbsp}

=== 8. Configure the Callback URL in Keycloak

After deployment, update the `eclipse-che` client in Keycloak with the redirect URI and web origin:

. Get the Eclipse Che URL:
+
[source,shell]
----
oc get checluster eclipse-che -n eclipse-che \
  -o jsonpath='{.status.cheURL}'
----
. In the Keycloak Admin Console, open the `eclipse-che` client.
. Paste the URL with `/oauth/callback` appended into *Valid redirect URIs*.
. Paste the URL into *Web origins*.
. Save.

// TODO: Add screenshot of Keycloak redirect URI configuration
// image::/assets/img/openshift-external-idp/keycloak-redirect-uri.png[Configuring the redirect URI in Keycloak]
{nbsp}

=== 9. Grant User Access

Create a `ClusterRoleBinding` so that users from your Keycloak group can access Che:

> [source,yaml]

Reviewer 2: **Inconsistent heading capitalization**: This heading uses lowercase "workspace" while all other headings use title case ("Grant User Access", "Deploy Eclipse Che", "Import the Keycloak CA Certificate").

```suggestion
=== 10. Start a Workspace
```

----
oc apply -f - <<EOF
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: openshift-developers

> roleRef:

Reviewer 2: **Future tense**: "you will land on", use present tense for immediate outcomes (IBM Style).

Suggestion:
```
After authenticating, the Eclipse Che dashboard opens.
```

  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: self-provisioner
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: Group
  name: <GROUP_NAME>
EOF
----

=== 10. Start a workspace

Once everything is configured, open the dashboard and start a workspace:

[source,shell]
----
chectl dashboard:open
----

You should be redirected to the Keycloak login page.
After authenticating, you will land on the Eclipse Che dashboard.

image::/assets/img/openshift-external-idp/dashboard.png[Eclipse Che dashboard after successful authentication]
{nbsp}

For the full reference documentation,

> see link:https://eclipse.dev/che/docs/stable/administration-guide/installing-che-on-openshift-with-keycloak-as-oidc/[Installing Che on OpenShift with Keycloak as external identity provider].

Reviewer 3: I noticed I get 404 error with `stable`,

```suggestion
see link:https://eclipse.dev/che/docs/next/administration-guide/installing-che-on-openshift-with-keycloak-as-oidc/[Installing Che on OpenShift with Keycloak as external identity provider].
```
